[FIX] contacts,*: improve contact computed fields

malh-odoo opened the pull request:
**Problem**:

The following res_partner computed fields are taking a lot of time to compute for contacts with 500K+ children (+2 minutes)
Fields:
    * total_invoiced
    * meeting_count
    * purchase_order_count
    * sale_order_count
    * supplier_invoice_count

The issue in four of thies five fields is comming from read method.
Prefetching using read was a necessary step before updating orm, as mentioned in the task related to the PR it was added in.
task link: https://www.odoo.com/odoo/all-tasks/1919933

**Solution**:
    * Avoid read method in computation of the fields.
    * Speedup sum generator expression of total_invoiced computation by replacing list with set.

**Benchmark**
| # of children | Before | After |
|--------|--------|--------|
| 8500 | 3s | 2.5s |
| 54000 | 30s | 15s |
| 580000 | 3.20m | 34s |

opw-4421866

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/194381.png)](https://mergebot.odoo.com/odoo/odoo/pull/194381)

ticodoo commented on addons/purchase/models/res_partner.py:
```diff
@@ -12,7 +12,6 @@ class res_partner(models.Model):
     def _compute_purchase_order_count(self):
         # retrieve all children partners and prefetch 'parent_id' on them
         all_partners = self.with_context(active_test=False).search([('id', 'child_of', self.ids)])
```
If you're going to remove the prefetch, then you should update the code comment to match.

Also, this code is super old, I don't understand why it needed to prefetch the `parent_id` in the first place or why it wasn't just using a `read_group` to only get the `id` considering it doesn't use `all_partners` for anything else.

All that being said, it seems like this issue was already fixed in 17+ via this: https://github.com/odoo/odoo/pull/112126/commits/83b9f7401ab32da1cb8ea5a99e04ad9f4d7ce0fe#diff-399ff553a811afbc017518d8ca7992de980422f6e9683504b4d3fe5183e22377 so it's maybe not worth changing in v16 for this edge case (500K+ children is a bit much), especially if the customer can just upgrade to v17 to fix the issue

ticodoo reviewed (commented):
Not sure this is worth it (see code comment). Also some general feedback:

Your commit message title is incorrect, we don't do [IMP] in stable, we only do [FIX] and on rare occasions `[PERF]`. Also you never put the model in the title, only the modules involved. Additionally, for performance PRs like this one, we typically expect more detailed benchmark info in the commit message or PR description with a time comparison of before + after for different amounts of records. E.g. (shamelessly stolen from another performance PR):

| # of children | Before | After |
|--------|--------|--------|
| 32000 | 24.08s | 1.43s |
| 7300 | 1.74s | 0.40s |

malh-odoo commented:
Hi @ticodoo 
Thank you for your comments.
I added the mentioned changes and I also added in the commit message the reason of using the "all_partners.read(['parent_id'])" in the first place (after asking the owner of the code).
Additionally, upgrading the client's database is not an easy option for them as they are using SH.
I look forward to your feedback.
Thank you

ticodoo reviewed (approved):
Hello @malh-odoo fair enough for the difficulty of upgrading. I'm still reluctant about the change since the task you pointed out explicitly added in the `read` for the speedup and I don't see where it says the prefetching was needed to update the orm, but since your change seems to have sped it up more, I guess the v10 ORM `search` just didn't prefetch as well as it does in v16 :shrug: 

Someone from @odoo/rd-subscription-sign-calendar still needs to review so I'll let it up to them to r+

@robodoo fw=no

robodoo commented:
Disabled forward-porting.

arj-odoo reviewed (commented):
There is a conflict and some useless diff. Otherwise, it looks ok :-).

malh-odoo commented:
@arj-odoo Can you check please

arj-odoo commented:
@robodoo r+
